### ISSUE TITLE
invalid mem access

### ISSUE THREAD

**Issue by iAklis** [iovisor/bcc#2471]
I got the invalid mem access

```
3: (85) call bpf_get_current_task#35
4: (79) r3 = *(u64 *)(r0 +1328)
R0 invalid mem access 'inv'
```

I change it from `ppid = task->real_parent->tgid;` and  follow the  https://github.com/iovisor/bcc/blob/master/docs/reference_guide.md#1-invalid-mem-access

now the c source is

```
  ...
    u64 ppid;
    struct task_struct* task = (struct task_struct*)bpf_get_current_task();
    bpf_probe_read(&ppid, sizeof(ppid), (void *)&(task->real_parent->tgid));
 ...
```
But didn't work.

**Comment by pchaigno** [comment]
Which version of bcc are you using?

You will need at least #1287 (v0.4.0) for `ppid = task->real_parent->tgid;` to work. 

> now the c source is
> 
> ```
>   ...
>     u64 ppid;
>     struct task_struct* task = (struct task_struct*)bpf_get_current_task();
>     bpf_probe_read(&ppid, sizeof(ppid), (void *)&(task->real_parent->tgid));
>  ...
> ```
> 
> But didn't work.

Could you try breaking that `bpf_probe_read` in two `bpf_probe_read`, one for `->real_parent`, then `->tgid`?

**Comment by iAklis** [comment]
@pchaigno Glad for your reply.

But didn't work.

3: (85) call bpf_get_current_task#35

Maybe the bpf_get_current_task is broken. 

I compiled my code which based on the tcptracer, because I want to get a elf.

**Comment by pchaigno** [comment]
> But didn't work.

Could you post the code that didn't work? What's the error message? Still the same? And which version of bcc are you using?

> Maybe the bpf_get_current_task is broken.

What makes you think that?

**Comment by yonghong-song** [comment]
```
3: (85) call bpf_get_current_task#35
4: (79) r3 = *(u64 *)(r0 +1328)
R0 invalid mem access 'inv'
```
insn #4 is not correct. We need to use bpf_probe_read() instead of a direct load.
as @pchaigno suggested, complete source code plus other info to reproduce your result is the best.

**Comment by iAklis** [comment]
@pchaigno I am sorry for my negligence.

Background:
At first I just wanted to compile ebpf as a separate binary. I got the Makefile, ebpf.mk and bpf_helpers.h from `weaveworks/tcptracer-bpf` as the my compile environment.

```
Maybe the bpf_get_current_task is broken.
```
The bpf_helpers.h I got from tcptracer-bpf has no  `bpf_get_current_task`, and I compile it without `bpf_get_current_task`, i got 
```
tcptracer-bpf.c:451:32: error: cast to 'struct task_struct *' from smaller integer type 'int' [-Werror,-Wint-to-pointer-cast]
    struct task_struct* task = (struct task_struct*)bpf_get_current_task();

```

So I copy it from https://github.com/iovisor/bcc/blob/a74c0429396f3180ba5b20f4a4eefc233b681cb4/src/cc/export/helpers.h#L403-L404

the tcptracer-ebpf.o I got successfully, but when I run it I got `invalid mem access`.

In the another one, the Message `R0 invalid mem access 'inv'` means that the return  value (pointer) of  bpf_get_current_task.


Apologize again, and 

**The way to reproduce:**
- I use `weaveworks/tcptracer-bpf`, modified the tcptracer-bpf.c
```c
SEC("kretprobe/tcp_v4_connect")
int kretprobe__tcp_v4_connect(struct pt_regs *ctx)
{
	int ret = PT_REGS_RC(ctx);
	u64 pid = bpf_get_current_pid_tgid();

    struct task_struct* task = (struct task_struct*)bpf_get_current_task();
    bpf_probe_read(&pid, sizeof(pid), &task->tgid);
/*Add two line above */
	struct sock **skpp;
	u64 zero = 0;
	struct tcptracer_status_t *status;

```

- add the follow to the bpf_helpers.h 

https://github.com/iovisor/bcc/blob/a74c0429396f3180ba5b20f4a4eefc233b681cb4/src/cc/export/helpers.h#L403-L404


- make(build docker and make the ebpf.mk)
- change the working directory to tests and go run tracer.go

Best Regards,
Aklis Chen
